### ISSUE TITLE
Sentinel: Modify muting transition to do nothing if the contact is already muted.

### ISSUE THREAD

**Issue by kitsao** [medic/cht-core#7020]
**Describe the bug**
When replicating muted contacts, a document gets the muted attribute updated to the date of replication. This is a source of conflicts.

**To Reproduce**
1. You need two instances, both running sentinel (muting transition)
2. Mute a contact in one instance (tested/observed in version 3.6.0)
3. Replicate the document in a destination (tested/observed in version 3.9.0) preferably on different dates.
4. Notice the document in the destination database gets muted updated to the replication date.

**Expected behavior**
The initial muting date should not be updated.

**Additional context**
Documents replicated from LG Mainstream instance (version 3.6.0) to CIHA instance (version 3.9.0) got the muted attribute updated to the date replication happened, causing conflicts. E.g. this document in the source
```
{
  "_id": "08a9977b-509e-47b4-8c75-03bf909c7726",
  "_rev": "6-f4ef913b23fbbaaa50a4a6d6a24dbc45",
  "name": "Some  Family",
  "type": "clinic",
  "reported_date": 1534867826795,
  "muted": "2020-12-07T07:05:45.405Z"
}
```
replicated to a destination on eb 25, would be updated to
```
{
  "_id": "08a9977b-509e-47b4-8c75-03bf909c7726",
  "_rev": "6-57a19f903726d51caf84487eb2abb1e8",
  "name": "Some  Family",
  "type": "clinic",
  "reported_date": 1534867826795,
  "muted": "2021-02-25T10:48:20.495Z"
}
```

**Comment by dianabarsan** [comment]
The muting transition already doesn't run if the main contact (the muting/unmuting report's subject) is already in the correct state: 
https://github.com/medic/cht-core/blob/master/shared-libs/transitions/src/transitions/muting.js#L121

but if's not (which is the case here, where the transitions runs on two servers at once ... ) and the muting/unmuting cascade starts (where descendants or ancestors are updated), indeed the code just updates regardless of current state: 
https://github.com/medic/cht-core/blob/0280e8caa08a2b7c65958b30d2ee8eb9dc1f7d60/shared-libs/transitions/src/lib/muting_utils.js#L58
So this is the place to change.

**Comment by dianabarsan** [comment]
I'm also sharing a comment I made on the Slack thread: 

If it's a race condition, where the muting report is replicated before the family (makes sense because it would be "first" in the changes feed), you might still be in this situation, even with this code change, because at the time where the second sentinel runs on the report, the family would not be muted on that server, because it wasn't replicated yet. 

You can even see proof of this, where the rev number is the same on both docs: 

1. both servers have the family are on rev 5
2. muting report happens
3. server1's sentinel processes the muting report and the family gets rev 6
4. replication starts
5. server2 receives report first (because it has precedence in the changes feed)
6. server2's sentinel processes the muting report and the family gets rev 6
7. replication reaches the family document
8. CONFLICT and both docs are on rev 6

**Comment by dianabarsan** [comment]
@kitsao 
I've just tested submitting additional muting reports against an already muted place, and the place was **not** updated by the followup reports. 

My configuration looks like: 
```
"muting": {
      "mute_forms": [
        "OFF"
      ],
      "unmute_forms": [
        "ON"
      ],
      "messages": [
        {
          "event_type": "already_muted",
          "recipient": "reporting_unit",
          "message": "ALREADY MUTED"
        }
      ]
    }
```
Where `ON` and `OFF` are taken from `standard` config. 

The second report got the `already_muted` message:
![image](https://user-images.githubusercontent.com/35681649/111581750-8170f180-87c2-11eb-8707-edb830123a72.png)

My place, and all it's descendants, are now on rev 3 (rev1 : creation, rev2: gets the patient_id/place_id, rev 3: gets the muted state by the 1st muting report) after submitting 
```
{
  "_id": "8d754948-c1a6-458a-9503-e5e28ab4ab38",
  "_rev": "3-0e6849caa6c0d7666352d91889ea060a",
  "parent": { ... },
  "type": "health_center",
  "name": "person area's Area",
  "reported_date": 1616047745648,
  "place_id": "20758",
  "muted": "2021-03-18T06:10:33.796Z"
}
{
  "_id": "0eb2676a-120e-440c-81a5-838ea58c9225",
  "_rev": "3-b695a79dddec9f0c82954c6a37865480",
  "parent": {
    "_id": "8d754948-c1a6-458a-9503-e5e28ab4ab38",
    "parent": {
      "_id": "8fe7d57e-a868-4efb-a06e-2cdf1bd138ff"
    }
  },
  "type": "clinic",
  "name": "family 1",
  "reported_date": 1616047766334,
  "place_id": "29270",
  "muted": "2021-03-18T06:10:33.796Z"
}
{
  "_id": "9d488d4d-f923-47ce-b13b-e49359dce04f",
  "_rev": "3-96bc56c4469193ba9507a068bbecac67",
  "type": "person",
  "name": "mom",
  "reported_date": 1616047766334,
  "parent": {
    "_id": "0eb2676a-120e-440c-81a5-838ea58c9225",
    "parent": {
      "_id": "8d754948-c1a6-458a-9503-e5e28ab4ab38",
      "parent": {
        "_id": "8fe7d57e-a868-4efb-a06e-2cdf1bd138ff"
      }
    }
  },
  "patient_id": "12452",
  "muted": "2021-03-18T06:10:33.796Z"
}
```

Can you, please, confirm this is the case?

**Comment by kitsao** [comment]
It's not the case @dianabarsan. In my case I have 3 instances, `LG mainstream` (3.6.0), `CIHA` (3.9.0) and `all-data` (3.6.0). `All-data` combines data from the two.
- Replication between `mainstream` and `CIHA` is as follows:
   - `mainstream` to `CIHA` is coninuous filtered by 20 CHW areas
   - `CIHA` to `mainstream` is continuous filtered by a set of reports e.g. muting, so that supervisors can verify for actual contact muting to happen. Supervisors are on `mainstream`.
- Replication between `mainstream`/`ciha` and `all-data`
   - `mainstream` to `all-data` is continuous, not filtered.
   - `ciha` to `all-data` is continuous filtered by doc types contacts and data_records.

##### That said,
- The update to muting date happens from `mainstream` to `CIHA`, on already muted contacts
   - This means we always have to deal with conflicts in `all-data`, which feeds into Klipfolio dashboards and causes discrepancies in indicators vs in-app analytics (CHWs are using CIHA).
   - The discrepancy can be explained in that the muted attribute gets excluded in the latest revision of the doc in `all-data`, meaning that some contacts muted in `mainstream` and `ciha` end up getting unmuted in `all-data`.

Happy to do a call to further clarify using some affected documents.

**Comment by dianabarsan** [comment]
Does the muting transition have to run on all instances? Can't you just keep the muting transition enabled on one instance? 
With continuous replication, there's a chance that both instances receive and process the muting report at the same time, causing the conflicts. The only way to guarantee no conflicts is to only run muting on one instance - the one that has all contacts (that doesn't get contact documents filtered from other instances).

**Comment by dianabarsan** [comment]
Another solution, if you absolutely need all / both servers to run muting, is to have different muting forms per server, and update the muting transition config so one server doesn't run muting on the same forms as the other.

**Comment by dianabarsan** [comment]
I'm not sure what is to be done here, since it's scheduled for 3.13. 
The issue here was that sentinel was running on two instances, where there's no guarantee that the muted contact will be synced before the muting report. So we could still be in a situation, even if we update the transition. 

I don't think there will be a way to avoid conflicts entirely when Sentinel is running on two instances that replicate against each other.

**Comment by dianabarsan** [comment]
I implemented what I described above, where muting won't change *secondary* (cascading) contacts unless necessary (when updating the muted state, or local muting history). 

I set up continuous replication between two databases, with Sentinel running on both, using the same config. I still got some conflicts. 

With the way that sentinel and replication work now, I'm not sure there's any chance we can overcome this and be able to support multiple Sentinel processes running in parallel over the same data, using the same configuration - even replicated data - without having to deal with conflicts. 

The change I made, to not update "secondary" contacts unless necessary, could still be a good change, albeit not entirely necessary and it won't fix the issue described above - it might reduce the number of conflicts, but it won't stop conflicts entirely. 

Thoughts?

**Comment by garethbowen** [comment]
It sounds like you've made an improvement so it'd be good to get that merged regardless, perhaps against a new issue which is more specific.

For the generic issue of multiple sentinels running I think we need a much bigger discussion about what problem that's trying to solve. I can see a case for horizontal scalability, but it'll be challenging to resolve this given the sequential nature of sentinel transitions. I've heard of situations where projects are running multiple versions of the CHT or config and trying to replicate the data, which is not a scenario we support, and would be extremely difficult to support well. If this is the case then I think we need to understand the actual use cases and attempt to find other solutions that are safer to maintain.

**Comment by dianabarsan** [comment]
I'm removing this from the workstream in favor of https://github.com/medic/cht-core/issues/7209 , which covers the changes in this comment: https://github.com/medic/cht-core/issues/7020#issuecomment-882686336 and some additional fixes.

**Comment by dianabarsan** [comment]
This cannot be reproducible in a "standard" CHT-Core deployment. The manifestation described here is entirely caused by the cht-to-cht replication between two instances that have the same config. 

In this case, it's recommended to only have one instance run Sentinel transitions instead of both.
